### ISSUE TITLE
Review Hosting API

### ISSUE THREAD

**Issue by rhuanjl** [microsoft/ChakraCore#6416]
The CC external API also called JSRT could do with a review prior to the next release.

Roughly speaking it contains:
1. "Windows only APIs" these are historic chakra.dll APIs that pre-date the cross platform nature of chakracore; they can actually be used on linux or macOS with some caution - notably they use strings that are rows of 16bit characters - normal on windows but not normal on other platforms. 

1. Settled Chakracore APIs:
Methods included in version 1.11 and in most cases several previous versions - many of these are marked as "experimental" in documentation but in practise have been stable for some years and should probably be labelled as such.

1. Master only APIs:
Various APIs have been added in master, some of these documented, some aren't - in particular several were added in #5923 that seem to follow a very different design philosophy to all the previous APIs. We should review these and consider removal or documentation on a case by case basis prior to a stable release.

**Comment by zenparsing** [comment]
Agreed - since we want CC to be an excellent engine for embedding we should give the API a lot of thought.

There are also some functions that are just really awkward (I'm thinking of `JsSetModuleHostInfo` in particular). Do you think we could get away with making some breaking changes just for `1.12`, to set us on a really solid path?

**Comment by Taritsyn** [comment]
> 1. "Windows only APIs":

In .NET I have long since given up using the APIs, that defined in the `ChakraCommonWindows.h` file. Usage of this APIs caused crashing of performance tests written by using the [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet) library:

```
// * Warnings *
Environment
  Summary -> Detected error exit code from one of the benchmarks. It might be caused by following antivirus software:
        - Windows Defender (windowsdefender://)
        - Norton Security (C:\Program Files\Norton Security\Engine\22.15.0.88\WSCStub.exe)
Use InProcessToolchain to avoid new process creation.
```

Having completely abandoned this APIs, nevertheless, some OS-specific usage remained. For example, on Windows I use the `JsCreateStringUtf16` and `JsCopyStringUtf16` methods to work with strings, and on Linux and Mac I use `JsCreateString` and `JsCopyString` methods. Unfortunately, now there is lack version of the `JsCreatePropertyId` method that would support UTF-16 encoding.

**Comment by rhuanjl** [comment]
> Agreed - since we want CC to be an excellent engine for embedding we should give the API a lot of thought.
> 
> There are also some functions that are just really awkward (I'm thinking of `JsSetModuleHostInfo` in particular). Do you think we could get away with making some breaking changes just for `1.12`, to set us on a really solid path?

I'm aware that the module API as a whole is not great - it has some awkward idiosyncrasies which I think stem from some missteps in its original implementation. `JsSetModuleHostInfo` probably has too many roles and the fact that it expects you to provide a moduleReference when half of its uses really need a scriptContext instead is silly. Additionally I think the whole logic of how ParseModuleSource works internally is unhelpful - which in turn means the callbacks you have to set up and the way the process works are not ideal.

This is probably the only part of the "settled" API I'd consider changing - though I'd be hesitant even here - could we perhaps introduce a better version and leave the existing one knocking around for backwards compatibility?

> > 1. "Windows only APIs":
> 
> In .NET I have long since given up using the APIs, that defined in the `ChakraCommonWindows.h` file. Usage of this APIs caused crashing of performance tests written by using the [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet) library:

I was not aware of this breakage - I wonder if we should entirely deprecate the "windows only" stuff - I'd like to keep it for backwards compatibility but I'm not keen on putting effort into fixing it. BUT there are a few things from it I could see trying to rescue - such as JsGetStringPointer which exposes a pointer to the internal string behind a Js String without copying it - there's potentially significant performance benefit to be had in using that in some cases.

> Having completely abandoned this APIs, nevertheless, some OS-specific usage remained. For example, on Windows I use the `JsCreateStringUtf16` and `JsCopyStringUtf16` methods to work with strings, and on Linux and Mac I use `JsCreateString` and `JsCopyString` methods. Unfortunately, now there is lack version of the `JsCreatePropertyId` method that would support UTF-16 encoding.

It sounds like a JsCreatePropertyIdUtf16 would be a sensible addition.

**Comment by rhuanjl** [comment]
@zenparsing I'm thinking more about that module API and weighing up my desire to avoid breakage against what a good API would actually look like:

1. The JsSetModuleHostInfo method is frankly horrid in part due to my own past additions to it, mea culpa, it has too many different purposes and the variety of activities that require in some cases a script context and in other cases a moduleRecord are confusing.
2. The basic mechanism for loading a module is unnecessarily abstruse and the interactions with the host are different to the spec.

| Activity        | ChakraCore           | Spec  |
|:------------- |:------------|:-----|
| List of Child Modules to fetch | Callback fired for each one | Method for Host to call to request list|
| Link module graph to prepare for execution | Done automatically when last module in graph is parsed | Host expected to track loading process and call a method when ready |
| Triggering execution | Notification via callback from link to trigger | As Link is synchronous host can call execution immediately after |

All that aside - whilst CC's method is initially hard to understand I think it actually requires the host to do a whole let less to implement it than the spec envisages. Notably per spec the host has to track/determine if the whole module graph is loaded whereas CC does that for you. 

I'd be interested what anyone else thinks of this and whether people would prefer:
a) Sticking to the existing API
b) Adding a new side by side API so maintaining backwards compatibility whilst adding a nicer alternative for new projects OR
c) Replacing the existing API (assume that not many existing projects use modules and hence go to town on making it nicer for future projects)

**Comment by rhuanjl** [comment]
Having another think on modules - with a possible path for alternate methods that could sit alongside the existing ones:
```c
// create a module record with the given normalizedSpecifier
// optionally a url can be supplied which will be used in error stack traces
// optionally hostDefined data can be supplied which can then be accessed via JsGetModuleRecordInfo
JsCreateModuleRecord(
    _In_ JsValueRef normalizedSpecifier,
    _In_opt_ JsValueRef url,
    _In_opt_ void* hostDefined,
    _Out_ JsModuleRecord moduleRecord);
```

```c
// retrieve data about a module record
JsGetModuleRecordInfo(
    _In_ JsModuleRecord moduleRecord,
    _Out_opt_ JsValueRef normalizedSpecifier,
    _Out_opt_ JsValueRef url,
    _Out_opt_ void* hostDefined);
```

```c
// set a module callback on the current context
// see JsModuleCallbackKind enum for callback types
JsSetModuleCallback(
    _In_ void* callback, 
    _In_ JsModuleCallbackKind callbackKind);
```

With these three APIs we'd eliminate the need for JsInitializeModuleRecord, JsSetModuleHostInfo and JsGetModuleHostInfo - BUT - we wouldn't break the mechanisms they sit on top of, so we could add these and leave the existing ones as deprecated but still functional for backwards compat.

**Comment by zenparsing** [comment]
@rhuanjl Yes, I think deprecating the old module API is the way to go.

- Do you think we can leave out `JsGetModuleRecordInfo` at first?
- Perhaps we should have separate functions for each callback type, so that the compiler can type check args?
- We'll also need a function for setting the module error, for instance when the module file can't be read.

**Comment by rhuanjl** [comment]
> @rhuanjl Yes, I think deprecating the old module API is the way to go.
> 
> * Do you think we can leave out `JsGetModuleRecordInfo` at first?
> * Perhaps we should have separate functions for each callback type, so that the compiler can type check args?
> * We'll also need a function for setting the module error, for instance when the module file can't be read.

1. I think you might need JsGetModuleRecordInfo in order to implement an import.meta callback. Also if we didn't have it hostDefined data on a module record would become meaningless.

2. I personally like having as few APIs as possible within reason - there are already 4 module callbacks to set AND top level await will add a 5th - I was tempted to propose one method that sets all of them but concerned a spec change could introduce a 6th later. That said I see the advantage/simplicity of an API per callback and hence not having an enum. I don't know what's best here.

3. We don't need a function to set a module error - JsParseModuleSource handles this if you call it with a nullptr as the sourceText parameter - this is actually what ch does.

**Comment by rhuanjl** [comment]
So see linked PR for merging of URL and Specifier to slightly tidy up what we need to do. I've also noticed that we need to be able to say if a module is a root module (currently done by providing a nullptr referencing module), so updated proposed API:

1. Module creation:
```c
// create a module record with the given normalizedSpecifier
// optionally a url can be supplied which will be used in error stack traces
// optionally hostDefined data can be supplied which can then be accessed via JsGetModuleRecordInfo
JsCreateModuleRecord(
    _In_opt_ JsValueRef normalizedSpecifier,
    _In_ bool isRootModule,
    _In_opt_ void* hostDefined,
    _Out_ JsModuleRecord moduleRecord);
```

2. Getting data (e.g. for `import.meta`)
```c
// retrieve data about a module record
JsGetModuleRecordInfo(
    _In_ JsModuleRecord moduleRecord,
    _Out_opt_ JsValueRef normalizedSpecifier,
    _Out_opt_ bool* isRootModule,
    _Out_opt_ void* hostDefined);
```

3. Setting callbacks - 5 variants of this, one for each callback (or one method with an enum OR one method that sets them all) I'm undecided.
```c
// set a module callback on the current context
JsSetFetchModuleCallback(
    _In_ FetchImportedModuleCallBack* callback);
```
